### PR TITLE
Aa/albedo

### PR DESCRIPTION
**Pull request by Aang23**
This is not fully-working & done code. 
It adds very basic Albedo support & custom renderers.
Once merged FavouriteDragon & I will continue working on it.

Do not merge until the upcoming bugfix is released.

### CODE REVIEW

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityFireArc.java [hunk 1]
```diff
@@ -375,4 +381,16 @@ public FireControlPoint(EntityArc arc, float size, double x, double y, double z)
...
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public Light provideLight() {
+		return Light.builder().pos(this).color(2F, 1F, 0F).radius(8).build();
+	}
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityFireArc.java]:
```suggestion
	public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityFireShot.java [hunk 2]
```diff
@@ -12,4 +18,16 @@ public EntityFireShot(World world) {
...
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public Light provideLight() {
+		return Light.builder().pos(this).color(2F, 1F, 0F).radius(8).build();
+	}
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityFireShot.java]:
```suggestion
	public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityFireball.java [hunk 3]
```diff
@@ -373,4 +380,16 @@ public void damageEntity(Entity entity) {
...
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public Light provideLight() {
+		return Light.builder().pos(this).color(2F, 1F, 0F).radius(10).build();
+	}
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityFireball.java]:
```suggestion
	public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityFlames.java [hunk 4]
```diff
@@ -197,4 +203,15 @@ public boolean isInRangeToRenderDist(double distance) {
...
 
+	@Override
+	@Optional.Method(modid = "albedo")
+	public Light provideLight() {
+		return Light.builder().pos(this).color(2F, 1F, 0F).radius(8).build();
+	}
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityFlames.java]:
```suggestion
	public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityLightOrb.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    @Override
+    @Optional.Method(modid = "albedo")
+    public Light provideLight() {
+        return Light.builder().pos(this).color(getColorR(), getColorG(), getColorB()).radius(getLightRadius()).build();
+    }
+
+    @Override
+    @Optional.Method(modid = "albedo")
+    public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityLightOrb.java]:
```suggestion
    public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java [hunk 6]
```diff
@@ -411,4 +418,21 @@ public void removeStatCtrl() {
 	public boolean isInRangeToRenderDist(double distance) {
 		return true;
 	}
+
+	@Override
+	public int getBrightnessForRender() {
+		return 15728880;
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java]: Just wondering, where does this number come from?

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java [hunk 7]
```diff
@@ -411,4 +418,21 @@ public void removeStatCtrl() {
...
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public Light provideLight() {
+		return Light.builder().pos(this).color(1F, 3F, 3F).radius(15).build();
+	}
+
+	@Override
+	@Optional.Method(modid = "albedo")
+	public void gatherLights(GatherLightsEvent arg0, Entity arg1) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java]:
```suggestion
	public void gatherLights(GatherLightsEvent event, Entity entity) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/data/LightOrbBehavior.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+	public static class Idle extends LightOrbBehavior {
+
+		@Override
+		public LightOrbBehavior onUpdate(EntityLightOrb entity) {
+			return this;
+		}
+
+		@Override
+		public void fromBytes(PacketBuffer buf) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/data/LightOrbBehavior.java]:
```suggestion
		public void fromBytes(PacketBuffer buffer) {
```

File: src/main/java/com/crowsofwar/avatar/common/entity/data/LightOrbBehavior.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		}
+
+		@Override
+		public void toBytes(PacketBuffer buf) {
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/common/entity/data/LightOrbBehavior.java]:
```suggestion
		public void toBytes(PacketBuffer buffer) {
```

File: src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+import net.minecraft.util.ResourceLocation;
+
+public class RenderLightOrb extends Render<EntityLightOrb> {
+
+    private ResourceLocation fill = new ResourceLocation("avatarmod", "textures/entity/sphere.png");
+    private ResourceLocation halo = new ResourceLocation("avatarmod", "textures/entity/spherehalo.png");
+
+    private CCModel model;
+
+    private CCModel cubeModel, sphereModel;
```
[the remark this answers is not in the document]
Comment (Aang23) [reply src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java]: They'll be if they end up being the same. Now I'm still unsure.

File: src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java (hunk 6 above)
Comment (Aang23) [reply src/main/java/com/crowsofwar/avatar/common/entity/EntityLightningSpear.java]: EntityFireArc

File: build.gradle [hunk 11]
```diff
@@ -103,6 +113,8 @@ minecraft {
 	replace "@name@", config.name
 }
 
+minecraft.coreMod = "elucent.albedo.asm.FMLPlugin"
```
[the remark this answers is not in the document]
Comment (Mahtaran) [reply build.gradle]: The `minecraft.` prefix is no longer necessary when fixed

[the remark this answers is not in the document]
Comment (Aang23) [reply build.gradle]: Oops... Did it too fast.

File: build.gradle [hunk 12]
```diff
@@ -96,6 +106,8 @@ minecraft {
 
 	runDir = "run"
 
+	minecraft.coreMod = "elucent.albedo.asm.FMLPlugin"
```
Comment (Mahtaran) [inline build.gradle]: The `minecraft.` prefix is no longer necessary when moved

File: src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        sphereModel = OBJParser.parseModels(new ResourceLocation("avatarmod", "models/hemisphere.obj")).get("model");
+    }
+
+    @Override
+    public void doRender(EntityLightOrb ent, double x, double y, double z, float entityYaw, float partialTicks) {
+
+        model = ent.isSphere() ? sphereModel : cubeModel;
+
+        Minecraft mc = Minecraft.getMinecraft();
+        Tessellator tess = Tessellator.getInstance();
```
[the remark this answers is not in the document]
Comment (Aang23) [reply src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java]: I actually do use it later l.87 "tessellator.draw();"

File: src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityFireball.java [hunk 14]
```diff
@@ -104,7 +105,6 @@ public void execute(AbilityContext ctx) {
 			data.addStatusControl(StatusControl.THROW_FIREBALL);
 			world.spawnEntity(fireball);
```
[the remark this answers is not in the document]
Comment (Aang23) [reply src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityFireball.java]: It was used during testing.

File: src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityFireball.java [hunk 15]
```diff
@@ -24,6 +24,7 @@
 import com.crowsofwar.avatar.common.data.BendingData;
 import com.crowsofwar.avatar.common.data.ctx.AbilityContext;
 import com.crowsofwar.avatar.common.entity.EntityFireball;
+import com.crowsofwar.avatar.common.entity.EntityLightOrb;
```
[the remark this answers is not in the document]
Comment (Aang23) [reply src/main/java/com/crowsofwar/avatar/common/bending/fire/AbilityFireball.java]: It was used during testing.

File: src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Mahtaran) [reply src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java]: Ah, totally missed that one.

**Comment by Mahtaran** [comment]
Now we just wait for @FavouriteDragon to review.

File: src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    private static ResourceLocation halo = new ResourceLocation("avatarmod", "textures/entity/spherehalo.png");
+    private ResourceLocation fill;
+
+    private CCModel model;
+
+    private static CCModel cubeModel, sphereModel;
+
+    public RenderLightOrb(RenderManager manager) {
+        super(manager);
+        cubeModel = OBJParser.parseModels(new ResourceLocation("avatarmod", "models/cube.obj")).get("model");
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java]: The assignments should be moved to the declaration.

File: src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+import net.minecraft.client.renderer.entity.RenderManager;
+import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
+import net.minecraft.util.ResourceLocation;
+
+/**
+ * @author Aang23
+ */
+public class RenderLightOrb extends Render<EntityLightOrb> {
+
+    private static ResourceLocation fill_def = new ResourceLocation("avatarmod", "textures/entity/sphere.png");
```
Comment (Mahtaran) [inline src/main/java/com/crowsofwar/avatar/client/render/RenderLightOrb.java]:
```suggestion
    private static ResourceLocation fill_default = new ResourceLocation("avatarmod", "textures/entity/sphere.png");
```

**Comment by FavouriteDragon** [comment]
@Mahtaran can we merge?

**Pull request merged**